packages/babel-helpers/src/helpers.js

```diff
@@ -270,6 +270,151 @@ helpers.createClass = template(`
   })()
 `);
 
+//NOTE: convention is 'descriptor' is used for element descriptors, while 'propertyDescriptor' is used for
+//property descriptor in all helpers related to decorators-2
+helpers.decorate = template(`
+  (function (constructor, undecorated, memberDecorators, heritage) {
+    const prototype = constructor.prototype;
+    let finishers = [];
+    const elementDescriptors = new Map(); // elementDescriptors is meant to be an array, so this will be converted later
+
+    for (const [key, isStatic] of undecorated) {
+      const target = isStatic ? constructor : prototype;
+      const propertyDescriptor = Object.getOwnPropertyDescriptor(target, key);
+      elementDescriptors.set(
+        [key, isStatic], 
+        babelHelpers.makeElementDescriptor(
+          "property",
+          key,
+          isStatic,
+          propertyDescriptor,
+        )
+      );
+    }
+
+    // decorate and store in elementDescriptors
+    for (const [key, decorators, isStatic] of memberDecorators) {
```

Comment:
Can we ES2015 here?

And a bit later, `Array.from` will require a polyfill.
It could cause issues like this https://github.com/babel/babel/issues/5876, right?